php_zookeeper.c

```diff
@@ -967,13 +1007,123 @@ static void php_zk_completion_marshal(int rc, const void *context)
 		php_error_docref(NULL TSRMLS_CC, E_WARNING, "could not invoke completion callback");
 	}
 
-#ifndef ZEND_ENGINE_3
+#ifdef ZEND_ENGINE_3
+#else
 	zval_ptr_dtor(&z_rc);
 #endif
 
 	if (cb_data->oneshot) {
-		zend_hash_index_del(&ZK_G(callbacks), cb_data->h);
+		zend_hash_index_del(cb_data->ht, cb_data->h);
+	}
+}
+
+static void php_zk_dispatch()
+{
+	TSRMLS_FETCH();
+
+	struct php_zk_pending_marshal *queue;
+	struct php_zk_pending_marshal *next;
+
+	if( !ZK_G(pending_marshals) ) {
+		return;
+	}
+
+	// Bail if the queue is empty or if we are already playing the queue
+	if( !ZK_G(head) || ZK_G(processing_marshal_queue) ) {
+		return;
+	}
+
+	// Prevent reentrant handler calls
+	ZK_G(processing_marshal_queue) = 1;
+
+	queue = ZK_G(head);
+	ZK_G(head) = NULL; /* simple stores are atomic */
+
+	while( queue ) {
+		// Process
+		if( queue->is_completion ) {
+			php_zk_dispatch_one_completion(queue->cb_data, queue->rc TSRMLS_CC);
+		} else {
+			php_zk_dispatch_one(queue->cb_data, queue->type, queue->state, queue->path TSRMLS_CC);
+		}
+
+		// Move
+		next = queue->next;
+		free(queue->path);
+		free(queue);
+		queue = next;
+	}
+}
+
+static void php_zk_watcher_marshal(zhandle_t *zk, int type, int state, const char *path, void *context)
+{
+	php_cb_data_t *cb_data = context;
+
+#if ZTS
+	void *prev = tsrm_set_interpreter_context(cb_data->ctx);
+	TSRMLS_FETCH_FROM_CTX(cb_data->ctx);
+#endif
+
+	// Allocate new item
+	struct php_zk_pending_marshal *p = calloc(1, sizeof(struct php_zk_pending_marshal));
+	p->cb_data = context;
+	p->type = type;
+	p->state = state;
+	p->path = strdup(path);
+	p->cb_data = cb_data;
+
+	// Add to list
+	if( ZK_G(head) && ZK_G(tail) ) {
```

Comment:
I not sure NTS vs ZTS matters too much here, I believe ZTS is mainly required for environments where PHP is being run within many threads in the same process. [[1]](https://jpauli.github.io/2017/01/12/threads-and-php.html#how-php-treats-requests)

I was considering adding a mutex to this block, but as there were none in pcntl, and portability may be an issue, I did not.

The pcntl extension may be a bit different, I guess signals probably use software interrupts instead of threads.

In any case, the conservative option would be to add a mutex here.